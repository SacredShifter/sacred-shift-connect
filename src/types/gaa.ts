export type ManifestDarkPhase = {
  duration: number;       // seconds
  intensity: number;      // 0..1
  curve: "linear" | "exp";
};

export type DarkEnergyParams = {
  driftRate: number;      // e.g., 0.0003
  depth: number;          // 0..1
};

export type GaaPolarity = {
  polarityEnabled: boolean;
  shadowMode: boolean;
  darkWeight: number;     // default 0.7
  lightWeight: number;    // default 0.3
  darkEnergyEnabled: boolean;
  darkEnergy: DarkEnergyParams;
  manifestDarkPhase: ManifestDarkPhase;
};

export type GaaParams = {
  // geometry
  R: number; r: number; n: number; phi0: number;
  omega: number; eta: number;                 // traversal & breath coupling
  kappaRef: number; tauRef: number;
  alpha: [number, number, number, number];    // freq sensitivities
  beta:  [number, number];                    // amp sensitivities
  gamma: [number, number];                    // filter sensitivities
  Lmin: number; Lmax: number;
};

export type EvidenceRef = {
  status: 'experimental'|'multi-source'|'peer-reviewed';
  refs: { title: string; doiOrUrl: string; summary: string }[];
};

export type GaaPreset = {
  id?: string;
  label: string;
  params: GaaParams;
  polarity: GaaPolarity;
  evidence?: EvidenceRef;
};

export type BioSignals = {
  breath: number;    // -1..1
  hrv?: number;      // 0..1 normalized
  eegBandRatio?: number; // 0..1
};

export type GaaCoreFrame = {
  // normalized core outputs from the geometric engine (NOT audio nodes)
  ThN: number; PhiN: number; kN: number; tN: number;
  dThNdt: number;             // normalized angular speed
  az: number; el: number;     // radians for spatialization
  f0: number; A0: number; fc0: number; // bases
};

export type GaaOutputs = {
  fHz: number;
  amp: number;       // 0..1
  fcHz: number;
  azimuth: number;
  elevation: number;
  darkPhaseActive: boolean;
  weights: { dark: number; light: number };
};

// ðŸ‘‡ This is what React components should receive (NOT the engine class)
export type ShadowEngineState = {
  t: number; // seconds elapsed
  darkPhaseActive: boolean;
  weights: { dark: number; light: number };
  lastOutputs: GaaOutputs;
};

// Simple biofeedback for UI display
export type SimpleBiofeedbackMetrics = {
  heartRateVariability: number;
  brainwaveActivity: {
    alpha: number;
    beta: number;
    theta: number;
    delta: number;
    gamma: number;
  };
  breathingPattern: {
    rate: number;
    depth: number;
    coherence: number;
  };
  autonomicBalance: {
    sympathetic: number;
    parasympathetic: number;
  };
};