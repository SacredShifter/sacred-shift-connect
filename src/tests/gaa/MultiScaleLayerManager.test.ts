import { describe, it, expect, vi } from 'vitest';
import { MultiScaleLayerManager } from '@/utils/gaa/MultiScaleLayerManager';
import { GeometricNormalizer } from '@/utils/gaa/GeometricNormalizer';

// Mock GeometricNormalizer since its implementation is not part of this test
vi.mock('@/utils/gaa/GeometricNormalizer', () => {
  return {
    GeometricNormalizer: vi.fn().mockImplementation(() => {
      return {
        normalize: vi.fn((geom) => ({ ...geom, normalized: true })),
      };
    }),
  };
});

describe('MultiScaleLayerManager', () => {
  it('should be instantiated without errors', () => {
    expect(() => new MultiScaleLayerManager()).not.toThrow();
  });

  it('should initialize with a default hierarchy of layers', () => {
    const manager = new MultiScaleLayerManager();
    const state = manager.getState();
    expect(state.layers).toBeDefined();
    expect(Object.keys(state.layers)).toEqual([
      'atomic',
      'molecular',
      'cellular',
      'tissue',
      'organ',
      'organism',
    ]);
    expect(state.layers.cellular.active).toBe(true);
  });

  it('should generate composite geometry without errors', () => {
    const manager = new MultiScaleLayerManager();
    let geometries;
    expect(() => {
      geometries = manager.generateCompositeGeometry(12);
    }).not.toThrow();

    expect(geometries).toBeInstanceOf(Array);
    expect(geometries.length).toBeGreaterThan(0);
    // @ts-expect-error - check for mock property
    expect(geometries[0].normalized).toBe(true);
  });

  it('should generate geometry with vertices', () => {
    const manager = new MultiScaleLayerManager();
    const geometries = manager.generateCompositeGeometry(12);

    expect(geometries[0].vertices).toBeInstanceOf(Array);
    expect(geometries[0].vertices.length).toBeGreaterThan(0);
    expect(geometries[0].vertices[0].length).toBe(3);
  });

  it('should update breath phase over time', () => {
    const manager = new MultiScaleLayerManager();
    const initialState = manager.getState();
    manager.updateBreathPhase(0.1);
    const updatedState = manager.getState();
    expect(updatedState.breathPhase).not.toEqual(initialState.breathPhase);
  });

  it('should toggle a layer on and off', () => {
    const manager = new MultiScaleLayerManager();
    const initialWeight = manager.getState().layers.organism.weight;

    manager.toggleLayer('organism');
    const stateAfterOff = manager.getState();
    expect(stateAfterOff.layers.organism.active).toBe(false);

    manager.toggleLayer('organism');
    const stateAfterOn = manager.getState();
    expect(stateAfterOn.layers.organism.active).toBe(true);
  });

  it('should not generate geometry for inactive layers', () => {
    const manager = new MultiScaleLayerManager();
    // Deactivate all but one layer
    manager.toggleLayer('atomic');
    manager.toggleLayer('molecular');
    manager.toggleLayer('cellular');
    manager.toggleLayer('tissue');
    manager.toggleLayer('organ');

    const geometries = manager.generateCompositeGeometry(20);
    // Only the 'organism' layer should produce geometry
    expect(geometries.length).toBe(1);
  });

  it('should produce a consistent geometric output for snapshot testing', () => {
    const manager = new MultiScaleLayerManager();
    // Mock performance.now() to ensure deterministic output for the snapshot
    vi.spyOn(performance, 'now').mockReturnValue(1000);

    const geometries = manager.generateCompositeGeometry(12);
    expect(geometries).toMatchSnapshot();
  });

  describe('Individual Layer Geometry Snapshots', () => {
    const layerIds: (keyof LayerHierarchy)[] = ['atomic', 'molecular', 'cellular', 'tissue', 'organ', 'organism'];

    layerIds.forEach(layerId => {
      it(`should generate a consistent snapshot for the ${layerId} layer`, () => {
        const manager = new MultiScaleLayerManager();
        vi.spyOn(performance, 'now').mockReturnValue(1000);

        // Deactivate all layers except the one being tested
        layerIds.forEach(id => {
          if (id !== layerId) {
            manager.toggleLayer(id);
          }
        });

        const geometry = manager.generateCompositeGeometry(20);
        expect(geometry).toMatchSnapshot();
      });
    });
  });
});
