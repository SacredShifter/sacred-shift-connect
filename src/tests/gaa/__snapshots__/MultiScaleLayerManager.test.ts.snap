// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiScaleLayerManager > Individual Layer Geometry Snapshots > should generate a consistent snapshot for the atomic layer 1`] = `
[
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        8,
        0,
        4,
      ],
      [
        9,
        13,
        4,
      ],
      [
        4,
        3,
        8,
      ],
      [
        8,
        5,
        0,
      ],
      [
        5,
        9,
        0,
      ],
      [
        0,
        9,
        4,
      ],
      [
        3,
        1,
        8,
      ],
      [
        8,
        6,
        5,
      ],
      [
        14,
        3,
        13,
      ],
      [
        13,
        3,
        4,
      ],
      [
        9,
        16,
        13,
      ],
      [
        5,
        12,
        9,
      ],
      [
        9,
        10,
        16,
      ],
      [
        12,
        10,
        9,
      ],
      [
        10,
        15,
        16,
      ],
      [
        16,
        15,
        13,
      ],
      [
        15,
        14,
        13,
      ],
      [
        7,
        6,
        8,
      ],
      [
        5,
        6,
        12,
      ],
      [
        1,
        7,
        8,
      ],
      [
        12,
        11,
        10,
      ],
      [
        6,
        11,
        12,
      ],
      [
        14,
        2,
        3,
      ],
      [
        3,
        2,
        1,
      ],
      [
        1,
        2,
        7,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.8003484062160007,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0,
        0,
        0.05,
      ],
      [
        0.07969056183313539,
        0.030872288126351025,
        0.046873450681232755,
      ],
      [
        0.09120242691548448,
        0.08314199770962895,
        0.005717894078934112,
      ],
      [
        0.03360985960908815,
        0.06749769715742403,
        -0.02847072446114903,
      ],
      [
        0.0012062152059209475,
        0.013017135142409706,
        -0.034966923976819365,
      ],
      [
        0.011494821094042188,
        -0.04040010557884155,
        -0.03391904307906029,
      ],
      [
        0.06424453388664614,
        -0.08507351171874576,
        -0.015166026621215118,
      ],
      [
        0.09884341264741726,
        -0.0612012500330349,
        0.028944471506810697,
      ],
      [
        0.041117021207750015,
        -0.0076861033363042446,
        0.05398686205844387,
      ],
      [
        -0.029901146861642587,
        -0.00558949305912451,
        0.039260360881991724,
      ],
      [
        -0.053373221269821795,
        -0.03304729948625677,
        0.015629364070843095,
      ],
      [
        -0.051681355460093115,
        -0.06843717485961548,
        -0.012200272385993175,
      ],
      [
        -0.017318178858584034,
        -0.06086708515043271,
        -0.05110267049390111,
      ],
      [
        -0.0021846381063005777,
        0.02357599981112868,
        -0.06333038599157084,
      ],
      [
        -0.037531517744761,
        0.07537344838861998,
        -0.03179273917677668,
      ],
      [
        -0.055968860177195184,
        0.05102235765036267,
        0.0035089418673960307,
      ],
      [
        -0.04594993392108617,
        0.017801099236430744,
        0.02702744104083316,
      ],
    ],
  },
]
`;

exports[`MultiScaleLayerManager > Individual Layer Geometry Snapshots > should generate a consistent snapshot for the cellular layer 1`] = `
[
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        2,
        1,
        17,
      ],
      [
        1,
        0,
        17,
      ],
      [
        0,
        16,
        17,
      ],
      [
        5,
        4,
        2,
      ],
      [
        2,
        3,
        1,
      ],
      [
        4,
        3,
        2,
      ],
      [
        5,
        2,
        17,
      ],
      [
        16,
        14,
        17,
      ],
      [
        8,
        5,
        17,
      ],
      [
        16,
        15,
        14,
      ],
      [
        15,
        13,
        14,
      ],
      [
        14,
        8,
        17,
      ],
      [
        11,
        8,
        14,
      ],
      [
        13,
        11,
        14,
      ],
      [
        7,
        6,
        5,
      ],
      [
        5,
        6,
        4,
      ],
      [
        8,
        7,
        5,
      ],
      [
        13,
        12,
        11,
      ],
      [
        12,
        10,
        11,
      ],
      [
        11,
        10,
        8,
      ],
      [
        8,
        9,
        7,
      ],
      [
        10,
        9,
        8,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.8881240871899965,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0.5530330085889911,
        0,
        0.07071067811865475,
      ],
      [
        0.48808708640308124,
        0.17764917118043763,
        0.09659258262890684,
      ],
      [
        0.3275265631824504,
        0.2748274183671562,
        0.0258819045102521,
      ],
      [
        0.2765165042944956,
        0.4789406345694039,
        -0.07071067811865475,
      ],
      [
        0.0901948479980313,
        0.5115204016744175,
        -0.09659258262890685,
      ],
      [
        -0.0742442443712261,
        0.42106003331378916,
        -0.02588190451025207,
      ],
      [
        -0.2765165042944954,
        0.47894063456940394,
        0.07071067811865474,
      ],
      [
        -0.39789223840505,
        0.33387123049397993,
        0.09659258262890685,
      ],
      [
        -0.4017708075536765,
        0.146232614946633,
        0.025881904510252254,
      ],
      [
        -0.5530330085889911,
        6.772701037913386e-17,
        -0.0707106781186548,
      ],
      [
        -0.4880870864030812,
        -0.17764917118043755,
        -0.09659258262890685,
      ],
      [
        -0.3275265631824505,
        -0.27482741836715624,
        -0.025881904510251925,
      ],
      [
        -0.27651650429449576,
        -0.47894063456940383,
        0.07071067811865478,
      ],
      [
        -0.09019484799803124,
        -0.5115204016744173,
        0.09659258262890685,
      ],
      [
        0.07424424437122597,
        -0.4210600333137892,
        0.025881904510252282,
      ],
      [
        0.2765165042944955,
        -0.4789406345694037,
        -0.07071067811865477,
      ],
      [
        0.39789223840505017,
        -0.3338712304939802,
        -0.09659258262890685,
      ],
      [
        0.4017708075536763,
        -0.1462326149466332,
        -0.025881904510252293,
      ],
    ],
  },
]
`;

exports[`MultiScaleLayerManager > Individual Layer Geometry Snapshots > should generate a consistent snapshot for the molecular layer 1`] = `
[
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        7,
        6,
        4,
      ],
      [
        6,
        5,
        4,
      ],
      [
        5,
        3,
        4,
      ],
      [
        4,
        1,
        7,
      ],
      [
        2,
        1,
        4,
      ],
      [
        1,
        9,
        7,
      ],
      [
        3,
        2,
        4,
      ],
      [
        9,
        8,
        7,
      ],
      [
        7,
        8,
        6,
      ],
      [
        1,
        10,
        9,
      ],
      [
        9,
        10,
        8,
      ],
      [
        2,
        0,
        1,
      ],
      [
        1,
        0,
        10,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.5152644553951775,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0.3048076211353316,
        0.15240381056766578,
        0.052794228634059935,
      ],
      [
        0.11148649022687246,
        0.18728112944254907,
        0.07265948095847506,
      ],
      [
        -0.030636538740042026,
        0.27785589290477436,
        0.03162417658202671,
      ],
      [
        -0.2024431063940061,
        0.2229386255539354,
        -0.07110547778631544,
      ],
      [
        -0.22839848195227255,
        0.0684588605833576,
        -0.0673135374260651,
      ],
      [
        -0.3390403305059032,
        -0.07123084425776155,
        -0.004980035943021929,
      ],
      [
        -0.20308793262070868,
        -0.18445691043538415,
        0.0784743851227857,
      ],
      [
        -0.04893947461032366,
        -0.21990299445953446,
        0.060980881160184824,
      ],
      [
        0.13377322598767755,
        -0.28938183816551477,
        -0.025457158406664925,
      ],
      [
        0.21416053893195355,
        -0.13207149761544806,
        -0.07744779788890857,
      ],
      [
        0.28831798854142077,
        -0.01189423411863881,
        -0.05022914500655629,
      ],
    ],
  },
]
`;

exports[`MultiScaleLayerManager > Individual Layer Geometry Snapshots > should generate a consistent snapshot for the organ layer 1`] = `
[
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        0,
        14,
        8,
      ],
      [
        8,
        7,
        0,
      ],
      [
        13,
        9,
        8,
      ],
      [
        14,
        13,
        8,
      ],
      [
        7,
        1,
        0,
      ],
      [
        12,
        10,
        9,
      ],
      [
        7,
        2,
        1,
      ],
      [
        13,
        12,
        9,
      ],
      [
        12,
        11,
        10,
      ],
      [
        6,
        2,
        7,
      ],
      [
        6,
        3,
        2,
      ],
      [
        5,
        3,
        6,
      ],
      [
        5,
        4,
        3,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.7293418736630729,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        1.125,
        0,
        0.54,
      ],
      [
        1.0099681050937124,
        0.44966677163442703,
        0.3300918620447381,
      ],
      [
        0.7029580936053614,
        0.7807140561585856,
        -0.037769324764344046,
      ],
      [
        0.29960088307027116,
        0.9220767055485415,
        -0.13500000000000004,
      ],
      [
        -0.09161722202344116,
        0.8716796406144484,
        0.055211903398969156,
      ],
      [
        -0.3937499999999999,
        0.6819950054802456,
        0.13500000000000004,
      ],
      [
        -0.5808508830702711,
        0.42201286898002116,
        -0.1349999999999999,
      ],
      [
        -0.6650589766756325,
        0.14136264938995588,
        -0.4825344406793634,
      ],
      [
        -0.6650589766756325,
        -0.1413626493899557,
        -0.4825344406793636,
      ],
      [
        -0.5808508830702712,
        -0.422012868980021,
        -0.13500000000000018,
      ],
      [
        -0.39375000000000027,
        -0.6819950054802452,
        0.1349999999999999,
      ],
      [
        -0.09161722202344191,
        -0.8716796406144479,
        0.055211903398969794,
      ],
      [
        0.29960088307027094,
        -0.9220767055485416,
        -0.13499999999999993,
      ],
      [
        0.7029580936053617,
        -0.7807140561585854,
        -0.03776932476434383,
      ],
      [
        1.0099681050937126,
        -0.44966677163442703,
        0.3300918620447381,
      ],
    ],
  },
]
`;

exports[`MultiScaleLayerManager > Individual Layer Geometry Snapshots > should generate a consistent snapshot for the organism layer 1`] = `
[
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        11,
        1,
        12,
      ],
      [
        1,
        13,
        12,
      ],
      [
        1,
        0,
        13,
      ],
      [
        11,
        3,
        1,
      ],
      [
        3,
        2,
        1,
      ],
      [
        11,
        6,
        3,
      ],
      [
        12,
        10,
        11,
      ],
      [
        11,
        10,
        9,
      ],
      [
        9,
        6,
        11,
      ],
      [
        5,
        4,
        3,
      ],
      [
        3,
        4,
        2,
      ],
      [
        9,
        7,
        6,
      ],
      [
        7,
        5,
        6,
      ],
      [
        6,
        5,
        3,
      ],
      [
        10,
        8,
        9,
      ],
      [
        9,
        8,
        7,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.6601287166998868,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        1,
        0,
        9.797174393178826e-17,
      ],
      [
        0.7566043134452798,
        0.419908788733975,
        -0.34710699129404643,
      ],
      [
        0.5655171470042862,
        0.660519905488284,
        -0.6254651859744238,
      ],
      [
        0.15977817235592529,
        0.6658533294698875,
        -0.779942329745459,
      ],
      [
        -0.14629594115857872,
        0.7304043519249112,
        -0.779942329745459,
      ],
      [
        -0.43767117122859206,
        0.48894156672894346,
        -0.625465185974424,
      ],
      [
        -0.5066462024446539,
        0.24817896462220895,
        -0.3471069912940467,
      ],
      [
        -0.5999999999999999,
        8.082668874372532e-17,
        -1.9594348786357652e-16,
      ],
      [
        -0.5925877755002428,
        -0.2481789646222089,
        0.3471069912940463,
      ],
      [
        -0.35991251174538164,
        -0.4889415667289433,
        0.6254651859744238,
      ],
      [
        -0.15956682941386147,
        -0.7304043519249112,
        0.7799423297454589,
      ],
      [
        0.1742720691392292,
        -0.6658533294698875,
        0.7799423297454591,
      ],
      [
        0.4650447874874745,
        -0.660519905488284,
        0.625465185974424,
      ],
      [
        0.8849458751985906,
        -0.4199087887339752,
        0.34710699129404676,
      ],
    ],
  },
]
`;

exports[`MultiScaleLayerManager > Individual Layer Geometry Snapshots > should generate a consistent snapshot for the tissue layer 1`] = `
[
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        12,
        11,
        10,
      ],
      [
        11,
        9,
        10,
      ],
      [
        10,
        3,
        12,
      ],
      [
        14,
        13,
        12,
      ],
      [
        12,
        13,
        11,
      ],
      [
        9,
        8,
        10,
      ],
      [
        16,
        14,
        1,
      ],
      [
        9,
        7,
        8,
      ],
      [
        8,
        5,
        10,
      ],
      [
        7,
        6,
        8,
      ],
      [
        1,
        14,
        12,
      ],
      [
        6,
        5,
        8,
      ],
      [
        16,
        15,
        14,
      ],
      [
        14,
        15,
        13,
      ],
      [
        5,
        3,
        10,
      ],
      [
        3,
        1,
        12,
      ],
      [
        5,
        4,
        3,
      ],
      [
        4,
        2,
        3,
      ],
      [
        3,
        2,
        1,
      ],
      [
        2,
        0,
        1,
      ],
      [
        1,
        0,
        16,
      ],
      [
        16,
        0,
        15,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.8125685356696833,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0.6587177826491071,
        0.3042487113059642,
        0.14,
      ],
      [
        0.40552043268640403,
        0.40315341508353725,
        -0.05990126329781921,
      ],
      [
        0.22551062585466525,
        0.5671031536786393,
        -0.22853513545974202,
      ],
      [
        -0.04105394953061298,
        0.5324948738753487,
        -0.27787774270814203,
      ],
      [
        -0.3014416471565431,
        0.5241631581024249,
        -0.18217312405718775,
      ],
      [
        -0.49636273830998473,
        0.38504439394042717,
        0.008622616395727622,
      ],
      [
        -0.6406654380910766,
        0.21693633703664653,
        0.19491750486961928,
      ],
      [
        -0.7167338663641066,
        0.017665867679885563,
        0.2794689320463723,
      ],
      [
        -0.6326140234399761,
        -0.17833966461815265,
        0.21814256086718778,
      ],
      [
        -0.5622969418383049,
        -0.38539722088760187,
        0.042949663366032106,
      ],
      [
        -0.30739072442277926,
        -0.45794743591603526,
        -0.15466219242894177,
      ],
      [
        -0.0931926015202694,
        -0.6020833952405577,
        -0.2715431420898027,
      ],
      [
        0.15501609972334396,
        -0.49307316859748324,
        -0.24668341440001962,
      ],
      [
        0.4439494996537426,
        -0.5017191034440466,
        -0.0930593438543048,
      ],
      [
        0.5251971779717792,
        -0.27840661701788644,
        0.10914004452195003,
      ],
      [
        0.7513639669666692,
        -0.13164099862062956,
        0.2543702761094665,
      ],
      [
        0.6264763451679426,
        0.0777976936395196,
        0.2668237601196038,
      ],
    ],
  },
]
`;

exports[`MultiScaleLayerManager > should produce a consistent geometric output for snapshot testing 1`] = `
[
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.1,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0,
        0,
        0.05,
      ],
      [
        2.4492935982947065e-17,
        -2.9995195653237156e-33,
        0.05,
      ],
    ],
  },
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.15,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0.3048076211353316,
        0.15240381056766578,
        0.052794228634059935,
      ],
      [
        -0.30480762113533166,
        -0.15240381056766572,
        0.05279422863405995,
      ],
    ],
  },
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        0,
        2,
        1,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.2,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0.5530330085889911,
        0,
        0.07071067811865475,
      ],
      [
        -0.2765165042944954,
        0.47894063456940394,
        0.07071067811865474,
      ],
      [
        -0.27651650429449576,
        -0.47894063456940383,
        0.07071067811865478,
      ],
    ],
  },
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        2,
        1,
        0,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.25,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        0.6587177826491071,
        0.3042487113059642,
        0.14,
      ],
      [
        -0.5537177826491068,
        0.2557512886940358,
        0.1399999999999998,
      ],
      [
        -1.2858791391047212e-16,
        -0.5600000000000003,
        -0.27999999999999997,
      ],
    ],
  },
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [
      [
        0,
        2,
        1,
      ],
    ],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.2,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        1.125,
        0,
        0.54,
      ],
      [
        -0.3937499999999999,
        0.6819950054802456,
        0.13500000000000004,
      ],
      [
        -0.39375000000000027,
        -0.6819950054802452,
        0.1349999999999999,
      ],
    ],
  },
  {
    "center": [
      0,
      0,
      0,
    ],
    "faces": [],
    "normalized": true,
    "normals": [
      [
        0,
        0,
        1,
      ],
      [
        0,
        0,
        1,
      ],
    ],
    "radius": 0.1,
    "sacredRatios": {
      "phi": 1.618033988749,
      "pi": 3.141592653589793,
      "sqrt2": 1.4142135623730951,
    },
    "vertices": [
      [
        1,
        0,
        9.797174393178826e-17,
      ],
      [
        -0.5999999999999999,
        8.082668874372532e-17,
        -1.9594348786357652e-16,
      ],
    ],
  },
]
`;
